### PR TITLE
Feat: add `DssLitePsmMom` + fix `IlkRegistry` class

### PR DESCRIPTION
**Pull request by amusingaxl**
Allows to set `tin` and `tout` to `type(uint256).max` to disable `sellGem` and/or `buyGem` for `DssLitePsm` without GSM delay.

### CODE REVIEW

File: src/DssLitePsmMom.sol [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+    function setAuthority(address authority_) external onlyOwner {
+        authority = authority_;
+        emit SetAuthority(authority_);
+    }
+
+    /*//////////////////////////////////
+      Governance actions without delay
+    //////////////////////////////////*/
+
+    function file(bytes32 key, bytes32 what, uint256 data) external auth {
```
Comment (sunbreak1211) [inline src/DssLitePsmMom.sol:123]: A few things about this function:
- I wouldn't reuse the `file` meme here. I think the purpose of the `PsmMom` will be to deactivate the Psm trading. So maybe it should be something like `disable` or so.
- I don't think we should allow a random value of `tin/tout`, the idea is having a safety method to disable the PSM without governance delay, not a way to randomly change the value of the fees without the delay (at least compared to other moms). So both values should be set at the same time to disable the module. Or maybe we could separate to disable inflows and outflows.
- The usage of `key` through `chainlog` is not bad but different from other existing `mom`s, where it passes directly the address instead of a `key` to look for the address.

Summarizing, I think we should be treating this module as a deactivation one and be separated from a possible future instant access module that could adjust `tin/tout` depending for example on an oracle price.

File: src/DssLitePsm.sol [hunk 2]
```diff
@@ -297,12 +299,15 @@ contract DssLitePsm {
...
      * @notice Function that swaps `gem` into Dai.
+     * @dev Reverts if `tin` is set to `type(uint256).max`.
      * @param usr The destination of the bought Dai.
      * @param gemAmt The amount of gem to sell. [`gem` precision].
      * @return daiOutWad The amount of Dai bought.
      */
     function sellGem(address usr, uint256 gemAmt) external returns (uint256 daiOutWad) {
-        daiOutWad = _sellGem(usr, gemAmt, tin);
+        uint256 tin_ = tin;
+        require(tin_ < type(uint256).max, "DssLitePsm/sell-gem-disabled");
```
Comment (sunbreak1211) [inline src/DssLitePsm.sol:309]: I think the idea to do this way was to avoid any extra check. Without the `require` it should always revert due to overflow/underflow (whatever is when multiplying by the fee or the in the next operation adding or substracting). I think this is one case where we should try to be a bit ingenuous and avoid changing the PSM (or at least the minimum as possible). We can leave comments to explain this.

Comment (amusingaxl) [reply src/DssLitePsm.sol:309]: I'm not sure about this one.
Technically, this is not a change in the behavior, since the functions will still revert. The only difference is that we are being explicit on why it is reverting, instead of relying on documentation to do so.
The gas cost impact is negligible here. We are introducing an extra `require` and caching the value into memory, but that allows us to keep the `unchecked` block in `_sellGem`, so only `buyGem*` functions would feel the impact.

File: src/DssLitePsmMom.sol (hunk 1 above)
Comment (amusingaxl) [reply src/DssLitePsmMom.sol]: Updated the code to reflect these suggestions.

File: src/DssLitePsm.sol (hunk 2 above)
Comment (oldchili) [reply src/DssLitePsm.sol:309]: Can we measure the gas consumption for both options in the 2 functions?
The difference is probably small, but this is the fast path.

Comment (amusingaxl) [reply src/DssLitePsm.sol:309]: The baseline is the current implementation and the delta is the version without the extra check and no `unchecked { }`:

```
Running 5 tests for src/DssLitePsm.integration.t.sol:DssLitePsmUsdcTest
[PASS] testBuyGem() (gas: 329297)
[PASS] testBuyGem_Revert_WhenPocketHasNoGem() (gas: 76259)
[PASS] testBuyGem_Revert_WhenToutIsMaxUint256() (gas: 35664)
[PASS] testSellGem() (gas: 260835)
[PASS] testSellGem_Revert_WhenTinIsMaxUint256() (gas: 35613)
Test result: ok. 5 passed; 0 failed; 0 skipped; finished in 25.22s
 
Ran 1 test suites: 5 tests passed, 0 failed, 0 skipped (5 total tests)
testSellGem() (gas: -34 (-0.013%)) 
testBuyGem() (gas: -68 (-0.021%)) 
testBuyGem_Revert_WhenPocketHasNoGem() (gas: -33 (-0.043%)) 
testBuyGem_Revert_WhenToutIsMaxUint256() (gas: 82 (0.230%)) 
testSellGem_Revert_WhenTinIsMaxUint256() (gas: 94 (0.265%)) 
Overall gas change: 41 (0.006%)
```

Comment (amusingaxl) [reply src/DssLitePsm.sol:309]: So yes, modifying the implementation allows us to save:
- 68 gas on `buyGem`: extra `require` with some `MLOAD`/`PUSH`/`POP` overhead.
- 34 gas on `sellGem`: the gain is partially offset because of the `unchecked {}` removal.

Considering `buyGem` uses around 330k gas and `sellGem` uses around 260k, we get a `0.02%` and a `0.01%` gas cost reduction respectively.

Honestly, it seems like too small of a gain to justify the loss of explicitness.

Comment (sunbreak1211) [reply src/DssLitePsm.sol:310]: I guess more than the gas efficiency what is bothering me is changing how clean the code looks until now.
I was feeling it was a clever solution making the mom to generate a revert situation as a disable feature and now we converted it as a highlighted special logic in the PSM itself.
I know we had to do something to the PSM anyway as the actual `file` wouldn't allow the max number, but in my mind was expecting to remove the restriction at all and removing the necessary unchecked/s to make that to work.
Anyway I guess it is more about a styling thing, I wasn't personally expecting the PSM to handle a disable logic as this case is probably just for some emergency situation that could allow to not have an explicit revert condition.

Comment (amusingaxl) [reply src/DssLitePsm.sol:310]: I believe the asymmetrical nature of `tin` and `tout` makes it weird to allow arbitrary values.

`sellGem` would revert for any `tin > 1 * WAD`, while `buyGem` can still work for a large range of values. For that reason, I decided to keep the existing constraint in place and introduce a special value to convey the different behavior.

Too much cleverness makes it harder to reason about the module. This is specially relevant in an emergency situation, as it would mean even more cognitive overload for people working at the protocol and for integrators.

We are still being a bit clever by reusing `tin` and `tout` storage slots to avoid extra gas costs, but we're being explicit to point out that the module will not work because of something extraordinary happened. To make it even more explicit, I'd also extract `type(uint256).max` into a well-named constant.

File: script/dependencies/DssLitePsmInit.integration.t.sol [hunk 3]
```diff
@@ -170,7 +188,7 @@ contract DssLitePsmInitTest is DssTest {
         // Global Line should be adjusted by increasing the new PSM line and reducing the old one
         {
             uint256 globalLine = dss.vat.Line();
-            assertEq(globalLine, pglobalLine - psrcLine + psrcArt * RAY + cfg.gap, "after: invalid Line change");
+            assertEq(globalLine, pglobalLine - psrcLine + ((psrcArt + cfg.buf) * RAY), "after: invalid Line change");
```
Comment (telome) [inline script/dependencies/DssLitePsmInit.integration.t.sol:191]: Not sure why but some of the deployment test fixes merged into `feat/implementations` are being reverted by this branch. Maybe due to having merged `feat/deploy-scripts` into this branch at some point before if was merged into `feat/implementations`?

Comment (amusingaxl) [reply script/dependencies/DssLitePsmInit.integration.t.sol:191]: Hmmm, good catch. Yeah, seem like a merge fuckup.

Comment (amusingaxl) [reply script/dependencies/DssLitePsmInit.integration.t.sol]: Fixed.

File: src/DssLitePsm.sol [hunk 4]
```diff
@@ -273,15 +275,17 @@ contract DssLitePsm {
      * @dev Swapping fees may not apply due to rounding errors for small swaps where
      *      `gemAmt < 10**gem.decimals() / tin` or
      *      `gemAmt < 10**gem.decimals() / tout`.
+     * @dev Setting `tin` or `tout` to `type(uint256).max` effectively halts `sellGem` and `buyGem` respectively,
+     *      due to overflow errors that would occur when trying to swap any value.
```
Comment (oldchili) [inline src/DssLitePsm.sol:279]: This is not correct currently, as we don't reach the overflow.

Comment (amusingaxl) [reply src/DssLitePsm.sol:279]: Yeah, that's why I wrote "would occur", but I should maybe rephrase that because it can look confusing.

File: src/DssLitePsmMom.t.sol [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+pragma solidity ^0.8.16;
+
```
Comment (oldchili) [inline src/DssLitePsmMom.t.sol:17]: Should this test file also be tagged as `integration`? it uses the chainlog and real contracts (as opposed to mocks).

File: src/DssLitePsmMom.t.sol [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+}
+
+contract InitCaller {
+    function init(DssInstance memory dss, DssLitePsmInstance memory inst, DssLitePsmInitConfig memory cfg) external {
+        DssLitePsmInit.init(dss, inst, cfg);
+    }
+}
+
+contract DssLitePsmMomTest is DssTest {
+    using stdStorage for StdStorage;
```
Comment (oldchili) [inline src/DssLitePsmMom.t.sol:43]: unused

File: src/DssLitePsmMom.t.sol [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+
+    DssLitePsm litePsm;
+    DssLitePsmMom mom;
+    ProxyLike pauseProxy;
+    address pause;
+    DssInstance dss;
+    ChiefLike chief;
+    DssLitePsmInstance inst;
+    DssLitePsmInitConfig cfg;
+    InitCaller caller = new InitCaller();
```
Comment (oldchili) [inline src/DssLitePsmMom.t.sol:60]: I think the initiation here is redundant.

File: src/DssLitePsmMom.t.sol [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+
+    function setUp() public {
+        vm.createSelectFork("mainnet");
+
+        dss = MCD.loadFromChainlog(CHAINLOG);
+        MCD.giveAdminAccess(dss);
```
Comment (oldchili) [inline src/DssLitePsmMom.t.sol:66]: unneeded?

File: src/DssLitePsmMom.sol [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+    //////////////////////////////////*/
+
+    /**
+     * @notice Halts either inflow or outflow of gems from the PSM.
+     * @param psm The PSM address.
+     * @param what The halted flow. [0 = `sellGem`, 1 = `buyGem`, 2 = `both`]
+     */
+    function halt(address psm, Flow what) external auth {
+        if (what == Flow.SELL || what == Flow.BOTH) {
+            DssLitePsmLike(psm).file("tin", DssLitePsmLike(psm).HALTED());
```
Comment (sunbreak1211) [inline src/DssLitePsmMom.sol:125]: nit: `HALTED()`could be saved in memory for avoiding double external call in `BOTH`. Anyway nothing that really matters if we prefer this way.

File: src/DssLitePsmMom.t.sol (hunk 6 above)
Comment (amusingaxl) [reply src/DssLitePsmMom.t.sol]: Fixed.

File: src/DssLitePsmMom.t.sol (hunk 7 above)
Comment (amusingaxl) [reply src/DssLitePsmMom.t.sol]: Fixed.

File: src/DssLitePsmMom.t.sol (hunk 8 above)
Comment (amusingaxl) [reply src/DssLitePsmMom.t.sol]: Removed.

File: src/DssLitePsmMom.sol (hunk 9 above)
Comment (amusingaxl) [reply src/DssLitePsmMom.sol]: Done.

File: src/DssLitePsmMom.t.sol (hunk 5 above)
Comment (amusingaxl) [reply src/DssLitePsmMom.t.sol]: Done.

File: src/DssLitePsm.sol (hunk 4 above)
Comment (amusingaxl) [reply src/DssLitePsm.sol]: Done.

File: src/DssLitePsmMom.integration.t.sol [hunk 10]
```diff
@@ -0,0 +1,169 @@
...
+
+        if (what == DssLitePsmMom.Flow.SELL || what == DssLitePsmMom.Flow.BOTH) {
+            assertEq(litePsm.tin(), litePsm.HALTED(), "doHalt: tin not set");
+        }
+        if (what == DssLitePsmMom.Flow.BUY || what == DssLitePsmMom.Flow.BOTH) {
+            assertEq(litePsm.tout(), litePsm.HALTED(), "doHalt: tout not set");
+        }
+    }
+
+    function testHaltFromOwner() public {
```
Comment (sunbreak1211) [inline src/DssLitePsmMom.integration.t.sol:154]: This test is a bit weird, as when you are checking `BOTH` it could actually be doing anything `halt` in that condition and you would still pass the asserts due to the prev `halt`s. I'd suggest separating the three of them in three different tests.

File: src/DssLitePsmMom.integration.t.sol [hunk 11]
```diff
@@ -0,0 +1,169 @@
...
+        doHalt(address(pauseProxy), DssLitePsmMom.Flow.SELL);
+        doHalt(address(pauseProxy), DssLitePsmMom.Flow.BUY);
+        doHalt(address(pauseProxy), DssLitePsmMom.Flow.BOTH);
+    }
+
+    function testHaltFromHat() public {
```
Comment (sunbreak1211) [inline src/DssLitePsmMom.integration.t.sol:160]: Same here.

File: script/dependencies/DssLitePsmInit.integration.t.sol [hunk 12]
```diff
@@ -52,35 +53,44 @@ contract InitCaller {
...
+    DssLitePsm litePsm;
+    DssLitePsmMom mom;
     InitCaller caller;
 
     function setUp() public {
         vm.createSelectFork("mainnet");
-        dss = MCD.loadFromChainlog(CHANGELOG);
+
+        dss = MCD.loadFromChainlog(CHAINLOG);
+        MCD.giveAdminAccess(dss);
```
Comment (sunbreak1211) [inline script/dependencies/DssLitePsmInit.integration.t.sol:84]: Shouldn't be needed here either.

File: src/DssLitePsmMom.integration.t.sol (hunk 10 above)
Comment (amusingaxl) [reply src/DssLitePsmMom.integration.t.sol:154]: Makes sense. I'll use `snapshot/revertTo` instead of breaking the tests.

Comment (amusingaxl) [reply src/DssLitePsmMom.integration.t.sol:158]: Done.

File: src/DssLitePsmMom.integration.t.sol (hunk 11 above)
Comment (amusingaxl) [reply src/DssLitePsmMom.integration.t.sol:164]: Done.

File: script/dependencies/DssLitePsmInit.integration.t.sol (hunk 12 above)
Comment (amusingaxl) [reply script/dependencies/DssLitePsmInit.integration.t.sol]: Removed.

**Review by oldchili — APPROVED** [review]
LGTM

**Review by sunbreak1211 — APPROVED** [review]
LGTM

**Review by telome — APPROVED** [review]
LGTM

File: script/dependencies/DssLitePsmInit.sol [hunk 13]
```diff
@@ -148,7 +151,7 @@ library DssLitePsmInit {
         require(src.ilk != ilk, "DssLitePsmInit/invalid-ilk-reuse");
 
         IlkRegistryLike reg = IlkRegistryLike(dss.chainlog.getAddress("ILK_REGISTRY"));
-        (src.name, src.symbol, src.class, src.dec, src.gem, src.pip, src.gemJoin,) = reg.info(src.ilk);
+        (src.name, src.symbol, , src.dec, src.gem, src.pip, src.gemJoin,) = reg.info(src.ilk);
```
Comment (sunbreak1211) [inline script/dependencies/DssLitePsmInit.sol:154]: We could remove the param from the `SrcPsm` struct.

Comment (amusingaxl) [reply script/dependencies/DssLitePsmInit.sol:153]: Done.

**Review by sunbreak1211 — APPROVED** [review]
LGTM

**Review by oldchili — APPROVED** [review]
LGTM

**Review by telome — APPROVED** [review]
LGTM

**Review by 0xp3th1um — APPROVED** [review]
lgtm
